compiler/rustc_codegen_llvm/src/back/write.rs

```diff
@@ -104,6 +104,33 @@ pub fn create_informational_target_machine(sess: &Session) -> &'static mut llvm:
         .unwrap_or_else(|err| llvm_err(sess.diagnostic(), err).raise())
 }
 
+pub fn create_target_machine_args(expanded_args: &[String]) -> *mut llvm::TargetMachineCppArgs {
+    let expanded_args_cstr: Vec<CString> =
+        expanded_args.iter().map(|s| CString::new(s.as_str()).ok()).flatten().collect();
+
+    let expanded_args_cstr_ptrs: Vec<*const c_char> =
+        expanded_args_cstr.iter().map(|cstr| cstr.as_ptr()).collect();
+
+    fn path_to_cstr(os: std::path::PathBuf) -> Option<CString> {
+        let s = os.to_str()?;
+        CString::new(s).ok()
+    }
+    let exe_path: CString =
+        std::env::current_exe().ok().map(path_to_cstr).flatten().unwrap_or_default();
+
+    let exe_path_ptr = exe_path.as_ptr();
+    let expanded_args_cstr_ptrs_slice = &expanded_args_cstr_ptrs[..];
+
+    // SAFETY: Ptrs are valid, and this function does make its own copies so we are allowed to invalidate the pointers after the functions exits
+    unsafe {
+        llvm::LLVMRustCreateTargetMachineArgs(
+            exe_path_ptr,
+            expanded_args_cstr_ptrs_slice.len() as c_int,
+            expanded_args_cstr_ptrs_slice.as_ptr(),
```

Comment:
```suggestion
            expanded_args_cstr_ptrs.as_ptr(),
```
`expanded_args_cstr_ptrs_slice` should be unnecessary, both `as_ptr` and `len` should work on regular vector.